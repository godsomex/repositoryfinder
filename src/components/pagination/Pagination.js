import React from "react";
import { PAGES_TO_SHOW } from "../../constants";

function getLowestPageToRender(totalPages, currentPage, numberOfPagesToShow) {
  if (numberOfPagesToShow >= totalPages) {
    return 1;
  }

  if (
    numberOfPagesToShow - currentPage >=
    Math.floor(numberOfPagesToShow / 2)
  ) {
    return 1;
  }
  const temp = currentPage - Math.floor(numberOfPagesToShow / 2);

  if (temp + (numberOfPagesToShow - 1) >= totalPages) {
    return totalPages - (numberOfPagesToShow - 1);
  }
  return temp;
}

function getPagesToRender(lowestPage, amount) {
  if (amount) {
    return new Array(amount)
      .fill(lowestPage)
      .map((value, index) => value + index);
  }
}

function Pagination({ reposPerPage, totalRepos, paginate, currentPage }) {
  const lastPage = Math.ceil(totalRepos / reposPerPage);
  const lowestPageToRender = getLowestPageToRender(
    lastPage,
    currentPage,
    PAGES_TO_SHOW
  );

  const actualPagesToShow = Math.min(
    PAGES_TO_SHOW,
    Math.ceil(totalRepos / reposPerPage)
  );
  const pagesToRender = getPagesToRender(lowestPageToRender, actualPagesToShow);

  if (totalRepos <= 25) return null;

  return (
    <nav className="d-flex justify-content-center">
      <ul className="pagination pagination-base pagination-boxed pagination-square mb-0">
        {currentPage !== 1 && (
          <li className="page-item">
            <a
              onClick={() => paginate(currentPage - 1)}
              className="page-link"
              href="!#"
            >
              Previous
            </a>
          </li>
        )}
        {pagesToRender?.map((number, index) => (
          <li
            className="page-item active"
            key={index}
            data-test="pagination-button"
          >
            <a
              onClick={() => paginate(number)}
              className="page-link no-border"
              href="!#"
            >
              {number}
            </a>
          </li>
        ))}
        {currentPage !== lastPage && (
          <li className="page-item">
            <a
              onClick={() => paginate(currentPage + 1)}
              className="page-link"
              href="!#"
            >
              Next
            </a>
          </li>
        )}
      </ul>
    </nav>
  );
}

export default Pagination;
