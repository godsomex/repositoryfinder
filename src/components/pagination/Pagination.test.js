import React from "react";
import { shallow, mount, render } from "enzyme";

import Pagination from "./Pagination";
describe("Pagination", () => {
  const props = {
    reposPerPage: 5,
    totalRepos: 1000,
    paginate: jest.fn(),
    currentPage: 1,
  };

  const wrapper = mount(<Pagination {...props} />);

  it("renders App without crashing ", () => {
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders 5 Pagination numbers ", () => {
    const element = wrapper.find('[data-test="pagination-button"]');
    expect(element.length).toBe(5);
    expect(wrapper).toMatchSnapshot();
  });
});
