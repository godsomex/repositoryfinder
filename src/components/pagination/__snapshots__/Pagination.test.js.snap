// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders 5 Pagination numbers  1`] = `ReactWrapper {}`;

exports[`Pagination renders App without crashing  1`] = `ReactWrapper {}`;
